provision.yml

```diff
@@ -7,147 +7,181 @@
-
-- hosts: selfsigned_certs
-  gather_facts: no
-  become: true
-  roles:
-    - { role: selfsigned_certs, tags: ['selfsigned_certs'] } 
-
-- hosts: php-apps-vm
-  become: true
-  roles:
-    - { role: hosts,     tags: ['hosts'] }
-    - { role: httpd,     tags: ['httpd'] }
-    - { role: welcome,   tags: ['welcome'] }
-
-- hosts: loadbalancer-vm
-  gather_facts: true
-  become: true
-  roles:
-    - { role: haproxy, tags: ['lb'] }
+    - role: common
+      tags: ['core', 'base', 'common']
+    - role: rsyslog
+      tags: ['core', 'base', 'rsyslog']
+    - role: iptables
+      when:
+        - iptables_enable | bool
+      tags: ['core', 'base', 'iptables']
+    - role: selfsigned_certs
+      when:
+        - use_selfsigned_certs
+      tags: ['core', 'base', 'selfsigned_certs']
 
-- hosts: loadbalancer_ha
-  gather_facts: true
+- hosts: frontend
   become: true
   roles:
-    - { role: keepalived, tags: ['keepalived'] }
-    - { role: bind, tags: ['bind'] }
+    - role: hosts
+      tags: ['core', 'frontend', 'hosts']
+    - role: httpd
+      tags: ['core', 'frontend', 'httpd']
+    - role: welcome
+      tags: ['core', 'frontend', 'welcome']
 
 - hosts: loadbalancer
   gather_facts: true
   become: true
   roles:
-    - { role: haproxy, tags: ['lb'] }
-
-- hosts: php-apps:java-apps:oidc:lifecycle
-  gather_facts: no
-  become: true
-  roles:
-    - { role: httpd,     tags: ['httpd'] }
-
-- hosts: storage
+    - role: haproxy
+      when:
+        - inventory_hostname not in groups['loadbalancer_ha']
+      tags: ['core', 'loadbalancer', 'lb']
+    - role: keepalived
+      when:
+        - inventory_hostname in groups['loadbalancer_ha']
+      tags: ['core', 'loadbalancer', 'keepalived']
+    - role: bind
+      when:
+        - inventory_hostname in groups['loadbalancer_ha']
+      tags: ['core', 'loadbalancer', 'bind']
+
+- hosts: db_mysql
   gather_facts: no
   become: true
   roles:
-    - { role: mysql, tags: ['mysql'] }
-
-- hosts: dbcluster
-  gather_facts: yes
-  become: true
-  roles:
-    - { role: galera, tags: ['galera'] }
-
-- hosts: dbcluster_nodes
-  gather_facts: yes
-  become: true
-  roles:
-    - { role: keepalived, tags: ['keepalived'] }
-
-- hosts: galera_provision_host
-  roles:
-    - { role: galera_create_users, tags: ['galera', 'galera_create_users'] }
-
-- hosts: mongo_servers
+    - role: mysql
+      when:
+        - inventory_hostname in groups['storage']
+      tags: ['core', 'db_mysql', 'mysql']
+    - role: galera
+      when:
+        - inventory_hostname in groups['dbcluster']
+      tags: ['core', 'db_mysql', 'galera']
+    - role: keepalived
+      when:
+        - inventory_hostname in groups['dbcluster_nodes']
+      tags: ['core', 'db_mysql', 'keepalived']
+    - role: galera_create_users
+      when:
+        - inventory_hostname in groups['galera_provision_host']
+      tags: ['core', 'db_mysql', 'galera', 'galera_create_users']
+
+- hosts: db_mongo
   gather_facts: yes
```

Comment:
Here the same comment: There is the default option of installing mysql, and the extra option to be high available, and install galera. The latter could be behind a feature toggle as well.

Reply:
Default `mysql` will be installed because the `storage` group has an host entry. The `galera` role will only be executed when there is an matching host entry in the `dbcluster` group.

Reply:
I see. So different concepts are used to deploy the right roles on the right hosts. Does it make sense to unify that ? I see these three methods now:

1) Feature toggles (like iptables_enable and selfsigned_certs)
2) Host groups (and using that to exclude roles from being run)
3) Tags (where I believe that the "core" tag is to deploy a minimal set right?)

Reply:
Yes, the "core" tag is the minimal set that will be deployed to have the basic set of services running.

In case of unifying it, will be difficult to answer that. The current way is mostly done to be backwards compatible, but also I don't have an exact way of how others are using it for their (production) environments. With for example for "mysql", a toggle `use_production_setup` can be used and if set to true, it will deploy the galera and keepalived.

But if you want to have seperate instances for running profile and engineblock, it might make more sense to have seperate groups for that and if the hostname is part of the group "profile", then "profile" will be installed on that host or not. If you want to use toggles for that, then you need to create/update the `host_vars/{{ inventory_hostname }}` but adding that toggle, or updating the inventory file and specifying the toggle over there (In which there are now 5 places (`template/vm/group_vars`, `template/vm/host_vars`, `template/vm/inventory`, `template/vm/secrets` and `group_vars`) where a property can be configured).

Also when using group, you can also make the "discovery" much more dynamic instead of using a property in group_vars that contains ip's or hostnames. With a `set_fact` task in a role you can create a new property and combine the hosts with an `,` seperator and use that in various configurations. So if a new host is added to that group, it will be automatically be picked up and configured on all places where you are using that property

With tags you just limit your ansible run and is always handy. If 1 role is updated, you only run that role. No toggle or group can fix that.